File: src/editor/render.c
#include "editor.h"

void editor_draw_rows(struct abuf *ab) {
    if (!EKO_COLOR_IS_NONE(config.color_bg)) {
        char buf[32];
        int len = eko_color_bg(config.color_bg, buf, sizeof(buf));
        abuf_append(ab, buf, len);
    }

    int gutter = editor.line_num_width;
    int text_cols = editor.win_size.cols - gutter;

    for (int y = 0; y < editor.win_size.rows; y++) {
        int filerow = y + editor.row_offset;
        if (filerow >= editor.rows) {
            if (gutter > 0) {
                char gpad[16];
                int glen = snprintf(gpad, sizeof(gpad), "%*s", gutter, "");
                abuf_append(ab, "\x1b[90m", 5);
                abuf_append(ab, gpad, glen);
                abuf_append(ab, "\x1b[39m", 5);
            }

            if (editor.rows == 0 && y == editor.win_size.rows / 3) {
                char welcome[sizeof(EKO_WELCOME)];
                int welcomelen = snprintf(welcome, sizeof(welcome), "%s", EKO_WELCOME);
                if (welcomelen > text_cols) welcomelen = text_cols;
                int padding = (text_cols - welcomelen) / 2;
                while (padding--) abuf_append(ab, " ", 1);
                abuf_append(ab, welcome, welcomelen);
            } else {
                abuf_append(ab, "~", 1);
            }
        } else {
            if (gutter > 0) {
                char lnum[16];
                int llen = snprintf(lnum, sizeof(lnum), "%*d ", gutter - 1, filerow + 1);
                abuf_append(ab, "\x1b[90m", 5);
                abuf_append(ab, lnum, llen);
                abuf_append(ab, "\x1b[39m", 5);
            }

            int len = editor.row[filerow].render_size - editor.col_offset;
            if (len < 0) len = 0;
            if (len > text_cols) len = text_cols;

            char *c = &editor.row[filerow].render[editor.col_offset];
            unsigned char *hl = editor.row[filerow].hl
                                ? &editor.row[filerow].hl[editor.col_offset]
                                : NULL;
            int current_color = EKO_COLOR_NONE;

            int sel_start_rx = -1, sel_end_rx = -1;
            if (editor.selecting) {
                int s_off, e_off;
                editor_get_sel_offsets(&s_off, &e_off);
                int row_off = editor.row[filerow].off;
                int row_end = row_off + editor.row[filerow].size;
                if (s_off < row_end && e_off > row_off) {
                    int sc = (s_off > row_off) ? s_off - row_off : 0;
                    int ec = (e_off < row_end) ? e_off - row_off : editor.row[filerow].size;
                    sel_start_rx = convert_row_to_render_x(&editor.row[filerow], sc)
                                   - editor.col_offset;
                    sel_end_rx   = convert_row_to_render_x(&editor.row[filerow], ec)
                                   - editor.col_offset;
                    if (sel_start_rx < 0) sel_start_rx = 0;
                    if (sel_end_rx > len) sel_end_rx = len;
                }
            }

            int in_sel = 0;
            for (int j = 0; j < len; j++) {
                int want_sel = (sel_start_rx >= 0 && j >= sel_start_rx && j < sel_end_rx);
                if (want_sel && !in_sel) {
                    char sbuf[32];
                    int slen = eko_color_bg(config.color_selection, sbuf, sizeof(sbuf));
                    if (slen > 0) abuf_append(ab, sbuf, slen);
                    else abuf_append(ab, "\x1b[7m", 4);
                    in_sel = 1;
                } else if (!want_sel && in_sel) {
                    if (!EKO_COLOR_IS_NONE(config.color_bg)) {
                        char bbuf[32];
                        int blen = eko_color_bg(config.color_bg, bbuf, sizeof(bbuf));
                        abuf_append(ab, bbuf, blen);
                    } else {
                        abuf_append(ab, "\x1b[49m", 5);
                    }
                    in_sel = 0;
                }

                if (iscntrl((unsigned char)c[j])) {
                    char sym = (c[j] <= 26) ? '@' + c[j] : '?';
                    if (!in_sel) abuf_append(ab, "\x1b[7m", 4);
                    abuf_append(ab, &sym, 1);
                    if (!in_sel) abuf_append(ab, "\x1b[m", 3);
                    if (!in_sel && !EKO_COLOR_IS_NONE(config.color_bg)) {
                        char buf[32];
                        int blen = eko_color_bg(config.color_bg, buf, sizeof(buf));
                        abuf_append(ab, buf, blen);
                    }
                    if (!EKO_COLOR_IS_NONE(current_color)) {
                        char buf[32];
                        int clen = eko_color_fg(current_color, buf, sizeof(buf));
                        abuf_append(ab, buf, clen);
                    }
                } else if (!hl || hl[j] == HL_NORMAL) {
                    if (!EKO_COLOR_IS_NONE(current_color)) {
                        abuf_append(ab, "\x1b[39m", 5);
                        current_color = EKO_COLOR_NONE;
                    }
                    abuf_append(ab, &c[j], 1);
                } else {
                    int color = editor_syntax_to_color(hl[j]);
                    if (color != current_color) {
                        current_color = color;
                        char buf[32];
                        int clen = eko_color_fg(color, buf, sizeof(buf));
                        abuf_append(ab, buf, clen);
                    }
                    abuf_append(ab, &c[j], 1);
                }
            }
            if (in_sel) {
                if (!EKO_COLOR_IS_NONE(config.color_bg)) {
                    char bbuf[32];
                    int blen = eko_color_bg(config.color_bg, bbuf, sizeof(bbuf));
                    abuf_append(ab, bbuf, blen);
                } else {
                    abuf_append(ab, "\x1b[49m", 5);
                }
            }
            abuf_append(ab, "\x1b[39m", 5);
        }

        abuf_append(ab, "\x1b[K", 3);
        abuf_append(ab, "\r\n", 2);
    }

    if (!EKO_COLOR_IS_NONE(config.color_bg))
        abuf_append(ab, "\x1b[49m", 5);
}

void editor_draw_status_bar(struct abuf *ab) {
    abuf_append(ab, "\x1b[7m", 4);
    char status[80], rstatus[80];

    int len = snprintf(status, sizeof(status), "%.20s%s - %d lines",
        file_name ? file_name : "[No Name]",
        editor.dirty ? " (modified)" : "",
        editor.rows);
    int rlen = snprintf(rstatus, sizeof(rstatus), "%s | %d/%d",
        editor.syntax ? editor.syntax->filetype : "no ft",
        editor.cpos.y + 1, editor.rows);

    if (len > editor.win_size.cols) len = editor.win_size.cols;
    abuf_append(ab, status, len);

    while (len < editor.win_size.cols) {
        if (editor.win_size.cols - len == rlen) {
            abuf_append(ab, rstatus, rlen);
            break;
        }
        abuf_append(ab, " ", 1);
        len++;
    }

    abuf_append(ab, "\x1b[m", 3);
    abuf_append(ab, "\r\n", 2);
}

void editor_draw_message_bar(struct abuf *ab) {
    abuf_append(ab, "\x1b[K", 3);
    int len = strlen(editor.status);
    if (len > editor.win_size.cols) len = editor.win_size.cols;
    if (len && time(NULL) - editor.status_time < 5)
        abuf_append(ab, editor.status, len);
}

void editor_refresh_screen(void) {
    editor_scroll();

    struct abuf ab = ABUF_INIT;
    abuf_append(&ab, "\x1b[?25l", 6);
    abuf_append(&ab, "\x1b[H", 3);

    editor_draw_rows(&ab);
    editor_draw_status_bar(&ab);
    editor_draw_message_bar(&ab);

    char buf[32];
    snprintf(buf, sizeof(buf), "\x1b[%d;%dH",
             (editor.cpos.y - editor.row_offset) + 1,
             (editor.render_x - editor.col_offset) + 1 + editor.line_num_width);
    abuf_append(&ab, buf, strlen(buf));
    abuf_append(&ab, "\x1b[?25h", 6);

    write(STDOUT_FILENO, ab.buf, ab.len);
    abuf_free(&ab);
}

File: src/editor/render.zig
const std = @import("std");
const terminal = @import("../terminal.zig");
const Editor = @import("Editor.zig");
const syntax_mod = @import("syntax.zig");
const Color = Editor.Color;
const HlType = Editor.HlType;

pub fn refreshScreen(self: *Editor) !void {
    self.scroll();

    self.render_buf.clearRetainingCapacity();
    const alloc = self.allocator;

    try self.render_buf.appendSlice(alloc, "\x1b[?25l");

    // Set background color if configured
    const has_bg = switch (self.config.color_bg) {
        .none => false,
        else => true,
    };
    if (has_bg) {
        try self.config.color_bg.writeBg(&self.render_buf, alloc);
    }

    try drawRows(self);
    try drawStatusBar(self);
    try drawMessageBar(self);

    if (has_bg) {
        try self.render_buf.appendSlice(alloc, "\x1b[49m");
    }

    // Position cursor (accounting for line number gutter)
    const cursor_row = self.cy - self.row_offset + 1;
    const cursor_col = self.rx - self.col_offset + 1 + self.line_num_width;
    var pos_buf: [32]u8 = undefined;
    const pos = std.fmt.bufPrint(&pos_buf, "\x1b[{d};{d}H", .{ cursor_row, cursor_col }) catch unreachable;
    try self.render_buf.appendSlice(alloc, pos);

    try self.render_buf.appendSlice(alloc, "\x1b[?25h");
    try terminal.write(self.render_buf.items);
}

fn drawRows(self: *Editor) !void {
    const alloc = self.allocator;
    const gutter = self.line_num_width;
    const text_cols = self.textCols();
    const sd = self.screen_dirty;

    for (0..self.screen_rows) |y| {
        if (!self.dirty_all) {
            if (sd) |d| {
                if (y < d.len and !d[y]) continue;
            }
        }

        var row_pos_buf: [16]u8 = undefined;
        const row_pos = std.fmt.bufPrint(&row_pos_buf, "\x1b[{d};1H", .{y + 1}) catch unreachable;
        try self.render_buf.appendSlice(alloc, row_pos);

        const file_row = y + self.row_offset;
        if (file_row < self.rows.items.len) {
            // Line number gutter
            if (gutter > 0) {
                try self.render_buf.appendSlice(alloc, "\x1b[90m");
                var lnum_buf: [16]u8 = undefined;
                const lnum = std.fmt.bufPrint(&lnum_buf, "{d}", .{file_row + 1}) catch "?";
                // Right-align the number in (gutter-1) chars + 1 space
                const num_width = gutter - 1;
                if (lnum.len < num_width) {
                    for (0..num_width - lnum.len) |_| try self.render_buf.append(alloc, ' ');
                }
                try self.render_buf.appendSlice(alloc, lnum);
                try self.render_buf.append(alloc, ' ');
                try self.render_buf.appendSlice(alloc, "\x1b[39m");
            }

            try appendVisibleRowWithSyntax(self, file_row, text_cols);
        } else {
            // Gutter for empty rows
            if (gutter > 0) {
                try self.render_buf.appendSlice(alloc, "\x1b[90m");
                for (0..gutter) |_| try self.render_buf.append(alloc, ' ');
                try self.render_buf.appendSlice(alloc, "\x1b[39m");
            }

            if (self.rows.items.len <= 1 and self.text.getTotalLength() == 0 and y == self.screen_rows / 3) {
                const welcome = "eko editor -- ^S=save ^F=find ^Q=quit";
                const wlen = @min(welcome.len, text_cols);
                const padding = if (text_cols > wlen) (text_cols - wlen) / 2 else 0;
                if (padding > 0) {
                    try self.render_buf.append(alloc, '~');
                    for (1..padding) |_| try self.render_buf.append(alloc, ' ');
                }
                try self.render_buf.appendSlice(alloc, welcome[0..wlen]);
            } else {
                try self.render_buf.append(alloc, '~');
            }
        }
        try self.render_buf.appendSlice(alloc, "\x1b[K");

        if (sd) |d| {
            if (y < d.len) d[y] = false;
        }
    }

    self.dirty_all = false;
}

fn appendVisibleRowWithSyntax(self: *Editor, row_idx: usize, max_cols: usize) !void {
    const alloc = self.allocator;
    const row = self.rows.items[row_idx];

    if (row.size == 0) return;

    // Read raw row content
    var stack_buf: [8192]u8 = undefined;
    var heap_buf: ?[]u8 = null;
    defer if (heap_buf) |hb| self.allocator.free(hb);

    const raw: []u8 = if (row.size <= stack_buf.len)
        stack_buf[0..row.size]
    else blk: {
        heap_buf = try self.allocator.alloc(u8, row.size);
        break :blk heap_buf.?;
    };
    self.text.copyRange(row.off, raw);

    // Expand tabs into rendered buffer
    const tab_size = self.config.tab_size;
    var rendered_storage: [16384]u8 = undefined;
    var rendered_len: usize = 0;
    for (raw) |c| {
        if (c == '\t') {
            const spaces = tab_size - (rendered_len % tab_size);
            for (0..spaces) |_| {
                if (rendered_len < rendered_storage.len) {
                    rendered_storage[rendered_len] = ' ';
                    rendered_len += 1;
                }
            }
        } else {
            if (rendered_len < rendered_storage.len) {
                rendered_storage[rendered_len] = c;
                rendered_len += 1;
            }
        }
    }
    const rendered = rendered_storage[0..rendered_len];

    // Compute syntax highlights
    var hl_storage: [16384]HlType = undefined;
    const hl = hl_storage[0..rendered_len];
    @memset(hl, .normal);

    if (self.syntax) |syn| {
        _ = syntax_mod.computeHighlights(syn, rendered, hl, row.hl_open_comment);
    }

    // Compute selection/flash range in rendered coordinates
    var sel_start_rx: ?usize = null;
    var sel_end_rx: ?usize = null;
    var sel_color = self.config.color_selection;

    // Determine absolute offsets for the highlighted region
    var s_off: ?usize = null;
    var e_off: ?usize = null;

    if (self.copy_flash_off) |flash| {
        // Copy flash takes priority over selection
        s_off = flash.start;
        e_off = flash.end;
        sel_color = self.config.color_copied;
    } else if (self.selection) |sel| {
        const a_off = if (sel.mark_y < self.rows.items.len)
            self.rows.items[sel.mark_y].off + sel.mark_x
        else
            self.text.getTotalLength();
        const b_off = self.cursorOffset();
        s_off = @min(a_off, b_off);
        e_off = @max(a_off, b_off);
    }

    if (s_off != null and e_off != null) {
        const row_off = row.off;
        const row_end = row_off + row.size;
        if (s_off.? < row_end and e_off.? > row_off) {
            const sc = if (s_off.? > row_off) s_off.? - row_off else 0;
            const ec = if (e_off.? < row_end) e_off.? - row_off else row.size;
            sel_start_rx = self.rowCxToRx(row_idx, sc);
            sel_end_rx = self.rowCxToRx(row_idx, ec);
            if (sel_start_rx.? < self.col_offset) sel_start_rx = self.col_offset;
        }
    }

    const col_start = self.col_offset;
    const col_end = col_start + max_cols;
    var current_color: Color = .none;
    var in_sel = false;

    var j = col_start;
    while (j < @min(col_end, rendered_len)) : (j += 1) {
        const want_sel = (sel_start_rx != null and sel_end_rx != null and
            j >= sel_start_rx.? and j < sel_end_rx.?);

        if (want_sel and !in_sel) {
            try sel_color.writeBg(&self.render_buf, alloc);
            in_sel = true;
        } else if (!want_sel and in_sel) {
            switch (self.config.color_bg) {
                .none => try self.render_buf.appendSlice(alloc, "\x1b[49m"),
                else => try self.config.color_bg.writeBg(&self.render_buf, alloc),
            }
            in_sel = false;
        }

        const color = syntax_mod.hlToColor(hl[j], &self.config);
        if (!color.eql(current_color)) {
            current_color = color;
            switch (color) {
                .none => try self.render_buf.appendSlice(alloc, "\x1b[39m"),
                else => try color.writeFg(&self.render_buf, alloc),
            }
        }

        try self.render_buf.append(alloc, rendered[j]);
    }

    if (in_sel) {
        switch (self.config.color_bg) {
            .none => try self.render_buf.appendSlice(alloc, "\x1b[49m"),
            else => try self.config.color_bg.writeBg(&self.render_buf, alloc),
        }
    }
    switch (current_color) {
        .none => {},
        else => try self.render_buf.appendSlice(alloc, "\x1b[39m"),
    }
}

fn drawStatusBar(self: *Editor) !void {
    const alloc = self.allocator;

    var sb_pos_buf: [16]u8 = undefined;
    const sb_pos = std.fmt.bufPrint(&sb_pos_buf, "\x1b[{d};1H", .{self.screen_rows + 1}) catch unreachable;
    try self.render_buf.appendSlice(alloc, sb_pos);
    try self.render_buf.appendSlice(alloc, "\x1b[7m");

    const name: []const u8 = if (self.filename) |f| f else "[No Name]";
    const modified: []const u8 = if (self.dirty > 0) " (modified)" else "";

    var status_buf: [256]u8 = undefined;
    const left = std.fmt.bufPrint(&status_buf, " {s}{s} - {d} lines", .{ name, modified, self.rows.items.len }) catch "[status]";
    const left_len = @min(left.len, self.screen_cols);
    try self.render_buf.appendSlice(alloc, left[0..left_len]);

    var right_buf: [64]u8 = undefined;
    const ft: []const u8 = if (self.syntax) |s| s.filetype else "no ft";
    const right = std.fmt.bufPrint(&right_buf, "{s} | {d}/{d} ", .{ ft, self.cy + 1, self.rows.items.len }) catch "";

    var col: usize = left_len;
    while (col < self.screen_cols) : (col += 1) {
        if (self.screen_cols - col == right.len) {
            try self.render_buf.appendSlice(alloc, right);
            break;
        }
        try self.render_buf.append(alloc, ' ');
    }
    try self.render_buf.appendSlice(alloc, "\x1b[m");
    try self.render_buf.appendSlice(alloc, "\r\n");
}

fn drawMessageBar(self: *Editor) !void {
    const alloc = self.allocator;

    var mb_pos_buf: [16]u8 = undefined;
    const mb_pos = std.fmt.bufPrint(&mb_pos_buf, "\x1b[{d};1H", .{self.screen_rows + 2}) catch unreachable;
    try self.render_buf.appendSlice(alloc, mb_pos);
    try self.render_buf.appendSlice(alloc, "\x1b[K");

    if (self.status_msg_len > 0) {
        const now = std.time.timestamp();
        if (now - self.status_time < 5) {
            const len = @min(self.status_msg_len, self.screen_cols);
            try self.render_buf.appendSlice(alloc, self.status_msg[0..len]);
        }
    }
}
